modules/repository/src/main/java/org/mule/runtime/module/repository/api/BundleDescriptor.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.repository.api;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+/**
+ * Descriptor to identify a bundle
+ *
+ * @since 4.0
+ */
+public class BundleDescriptor
+{
+
+    private String groupId;
+    private String artifactId;
+    private String version;
+    private String type;
+
+    private BundleDescriptor()
+    {
+    }
+
+    public String getGroupId()
+    {
+        return this.groupId;
+    }
+
+    public String getArtifactId()
+    {
+        return this.artifactId;
+    }
+
+    public String getVersion()
+    {
+        return this.version;
+    }
+
+    public String getType()
+    {
+        return type;
+    }
+
+    public String toString()
+    {
+        return format("BundleDescriptor { %s:%s:%s:$s }", groupId, artifactId, version, type);
+    }
+
+    /**
+     * Builder for creating a {@code BundleDescriptor}
+     */
+    public static class Builder
+    {
+
+        private static final String REQUIRED_FIELD_NOT_FOUND_TEMPLATE = "bundle cannot be created with null %s";
+
+        private BundleDescriptor bundleDescriptor = new BundleDescriptor();
+
+        /**
+         * @param groupId the group id of the bundle
```

Comment:
Can you document in each method whether or not the parameter can be null/empty and add the check?